### ISSUE TITLE
[FEATURE] - ADD ABILITY FOR COMMITING BRANCH/PR TO GIT REPO

### ISSUE THREAD

**Issue by patrick-hermann-sva** [stuttgart-things/machineshop#44]
WE WANT TO ADD A FEATURE FOR CREATING FILES, BRANCHES + PRS WITH ONE COMMAND OF MACHINESHOP 

### EASIEST PR CREATION (FILES+BRANCH+PR)

```bash
machineshop push --target pr \
--files "./lint.yaml:.github/workflows/call-ansible-lint.yaml, ./lint-profile.yaml:ansible-lint" \ # SOURCE(LOCAL):TARGET(BRANCH-REPO)
--repo machineshop \
--branch update-ansible-workflows \
--title added-lint-profile 
```

### FLAGS + DEFAULTS
-> --repo NO DEFAULT / VALUE MANDATORY - NOT EMPTY
-> --title NO DEFAULT / VALUE MANDATORY - NOT EMPTY
-> --files NO DEFAULT / NOT MANDATORY
-> --owner [stuttgart-things] 
-> --authorName=[os.Getenv("GITHUB_AUTHOR") machineShop] 
-> --authorEmail=[os.Getenv("GITHUB_EMAIL") or machine-shop@stuttgart-things.com] 
-> --token=[os.Getenv("GITHUB_TOKEN")] 
-> --description=[title] 
-> --commitMessage=[title] 
-> --baseBranch[main] 

* How to auto merge (after pr was created)?
* How to set labels on a pr?

machineShop subcommand code / functions to -> [machineshop](https://github.com/stuttgart-things/machineshop/blob/main/cmd/push.go)
github code / functions to -> [sthings-cli](https://github.com/stuttgart-things/sthingsCli) github.go

**Comment by patrick-hermann-sva** [comment]
snippet w/ google github package

```go




```go
package main

import (
	"bytes"
	"context"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
	"time"

	"github.com/ProtonMail/go-crypto/openpgp"
	"github.com/google/go-github/v61/github"
)

var (
	sourceOwner   = flag.String("source-owner", "stuttgart-things", "Name of the owner (user or org) of the repo to create the commit in.")
	sourceRepo    = flag.String("source-repo", "machineshop", "Name of repo to create the commit in.")
	commitMessage = flag.String("commit-message", "test", "Content of the commit message.")
	commitBranch  = flag.String("commit-branch", "test", "Name of branch to create the commit in. If it does not already exists, it will be created using the `base-branch` parameter")
	repoBranch    = flag.String("repo-branch", "machineshop", "")
	baseBranch    = flag.String("base-branch", "main", "")
	prRepoOwner   = flag.String("merge-repo-owner", "stuttgart-things", "")
	prRepo        = flag.String("merge-repo", "machineshop", "")
	prBranch      = flag.String("merge-branch", "main", "")
	prSubject     = flag.String("pr-title", "test", "")
	prDescription = flag.String("pr-text", "test", "")
	sourceFiles   = flag.String("files", "go.sum", "")
	authorName    = flag.String("author-name", "patrick-hermann-sva.de", "")
	authorEmail   = flag.String("author-email", "patrick-hermann-sva.de", "")
	privateKey    = flag.String("private-key", "", "")
)

var client *github.Client
var ctx = context.Background()

// getRef returns the commit branch reference object if it exists or creates it
// from the base branch before returning it.
func getRef() (ref *github.Reference, err error) {
	if ref, _, err = client.Git.GetRef(ctx, *sourceOwner, *sourceRepo, "refs/heads/"+*commitBranch); err == nil {
		return ref, nil
	}

	// We consider that an error means the branch has not been found and needs to
	// be created.
	if *commitBranch == *baseBranch {
		return nil, errors.New("the commit branch does not exist but `-base-branch` is the same as `-commit-branch`")
	}

	if *baseBranch == "" {
		return nil, errors.New("the `-base-branch` should not be set to an empty string when the branch specified by `-commit-branch` does not exists")
	}

	var baseRef *github.Reference
	if baseRef, _, err = client.Git.GetRef(ctx, *sourceOwner, *sourceRepo, "refs/heads/"+*baseBranch); err != nil {
		return nil, err
	}
	newRef := &github.Reference{Ref: github.String("refs/heads/" + *commitBranch), Object: &github.GitObject{SHA: baseRef.Object.SHA}}
	ref, _, err = client.Git.CreateRef(ctx, *sourceOwner, *sourceRepo, newRef)
	return ref, err
}

// getTree generates the tree to commit based on the given files and the commit
// of the ref you got in getRef.
func getTree(ref *github.Reference) (tree *github.Tree, err error) {
	// Create a tree with what to commit.
	entries := []*github.TreeEntry{}

	// Load each file into the tree.
	for _, fileArg := range strings.Split(*sourceFiles, ",") {
		file, content, err := getFileContent(fileArg)
		if err != nil {
			return nil, err
		}
		entries = append(entries, &github.TreeEntry{Path: github.String(file), Type: github.String("blob"), Content: github.String(string(content)), Mode: github.String("100644")})
	}

	tree, _, err = client.Git.CreateTree(ctx, *sourceOwner, *sourceRepo, *ref.Object.SHA, entries)
	return tree, err
}

// getFileContent loads the local content of a file and return the target name
// of the file in the target repository and its contents.
func getFileContent(fileArg string) (targetName string, b []byte, err error) {
	var localFile string
	files := strings.Split(fileArg, ":")
	switch {
	case len(files) < 1:
		return "", nil, errors.New("empty `-files` parameter")
	case len(files) == 1:
		localFile = files[0]
		targetName = files[0]
	default:
		localFile = files[0]
		targetName = files[1]
	}

	b, err = os.ReadFile(localFile)
	return targetName, b, err
}

// pushCommit creates the commit in the given reference using the given tree.
func pushCommit(ref *github.Reference, tree *github.Tree) (err error) {
	// Get the parent commit to attach the commit to.
	parent, _, err := client.Repositories.GetCommit(ctx, *sourceOwner, *sourceRepo, *ref.Object.SHA, nil)
	if err != nil {
		return err
	}
	// This is not always populated, but is needed.
	parent.Commit.SHA = parent.SHA

	// Create the commit using the tree.
	date := time.Now()
	author := &github.CommitAuthor{Date: &github.Timestamp{Time: date}, Name: authorName, Email: authorEmail}
	commit := &github.Commit{Author: author, Message: commitMessage, Tree: tree, Parents: []*github.Commit{parent.Commit}}
	opts := github.CreateCommitOptions{}
	if *privateKey != "" {
		armoredBlock, e := os.ReadFile(*privateKey)
		if e != nil {
			return e
		}
		keyring, e := openpgp.ReadArmoredKeyRing(bytes.NewReader(armoredBlock))
		if e != nil {
			return e
		}
		if len(keyring) != 1 {
			return errors.New("expected exactly one key in the keyring")
		}
		key := keyring[0]
		opts.Signer = github.MessageSignerFunc(func(w io.Writer, r io.Reader) error {
			return openpgp.ArmoredDetachSign(w, key, r, nil)
		})
	}
	newCommit, _, err := client.Git.CreateCommit(ctx, *sourceOwner, *sourceRepo, commit, &opts)
	if err != nil {
		return err
	}

	// Attach the commit to the master branch.
	ref.Object.SHA = newCommit.SHA
	_, _, err = client.Git.UpdateRef(ctx, *sourceOwner, *sourceRepo, ref, false)
	return err
}

// createPR creates a pull request. Based on: https://godoc.org/github.com/google/go-github/github#example-PullRequestsService-Create
func createPR() (err error) {
	if *prSubject == "" {
		return errors.New("missing `-pr-title` flag; skipping PR creation")
	}

	if *prRepoOwner != "" && *prRepoOwner != *sourceOwner {
		*commitBranch = fmt.Sprintf("%s:%s", *sourceOwner, *commitBranch)
	} else {
		prRepoOwner = sourceOwner
	}

	if *prRepo == "" {
		prRepo = sourceRepo
	}

	newPR := &github.NewPullRequest{
		Title:               prSubject,
		Head:                commitBranch,
		HeadRepo:            repoBranch,
		Base:                prBranch,
		Body:                prDescription,
		MaintainerCanModify: github.Bool(true),
	}

	pr, _, err := client.PullRequests.Create(ctx, *prRepoOwner, *prRepo, newPR)
	if err != nil {
		return err
	}

	fmt.Printf("PR created: %s\n", pr.GetHTMLURL())
	return nil
}

func main() {
	flag.Parse()
	token := os.Getenv("GITHUB_TOKEN")
	if token == "" {
		log.Fatal("Unauthorized: No token present")
	}
	if *sourceOwner == "" || *sourceRepo == "" || *commitBranch == "" || *sourceFiles == "" || *authorName == "" || *authorEmail == "" {
		log.Fatal("You need to specify a non-empty value for the flags `-source-owner`, `-source-repo`, `-commit-branch`, `-files`, `-author-name` and `-author-email`")
	}
	client = github.NewClient(nil).WithAuthToken(token)

	ref, err := getRef()
	if err != nil {
		log.Fatalf("Unable to get/create the commit reference: %s\n", err)
	}
	if ref == nil {
		log.Fatalf("No error where returned but the reference is nil")
	}

	tree, err := getTree(ref)
	if err != nil {
		log.Fatalf("Unable to create the tree based on the provided files: %s\n", err)
	}

	if err := pushCommit(ref, tree); err != nil {
		log.Fatalf("Unable to create the commit: %s\n", err)
	}

	if err := createPR(); err != nil {
		log.Fatalf("Error while creating the pull request: %s", err)
	}
}
```
```

**Comment by patrick-hermann-sva** [comment]
implemented - release 1.8.0
